perf_bench_test.go

```diff
@@ -24,127 +24,567 @@ package main
+	profile.CSVPlotter(*vmIncrStep, *benchDir, metrFile)
+
+	tearDownVMs(t, images, startVMID, isSyncOffload)
+}
+
+func TestProfileSingleConfiguration(t *testing.T) {
 	var (
 		servedTh      uint64
 		pinnedFuncNum int
-		vmID          int
 		isSyncOffload bool = true
-		images             = getImages()
-		funcs              = []string{}
-		timeInterval       = time.Duration(time.Second.Nanoseconds() / int64(*targetReqPerSec))
-		totalRequests      = *executionTime * *targetReqPerSec
+		images             = getImages(t)
 	)
 
 	log.SetLevel(log.InfoLevel)
 
-	funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
+	checkInputValidation(t)
 
 	createResultsDir()
 
-	// Pull images
-	for funcName, imageName := range images {
-		resp, _, err := funcPool.Serve(context.Background(), "plr_fnc", imageName, "record")
+	funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
 
-		require.NoError(t, err, "Function returned error")
-		require.Equal(t, resp.Payload, "Hello, record_response!")
+	plrfncPid := pullImages(t, images)
+
+	bootVMs(t, images, 0, *vmNum, plrfncPid)
+
+	serveMetrics := loadAndProfile(t, images, *vmNum, *targetRPS, isSyncOffload)
 
-		// For future use
-		// createSnapshots(t, concurrency, vmID, imageName, isSyncOffload)
-		// log.Info("Snapshot created")
-		// createRecords(t, concurrency, vmID, imageName, isSyncOffload)
-		// log.Info("Record done")
+	tearDownVMs(t, images, *vmNum, isSyncOffload)
+
+	dumpMetrics(t, []map[string]float64{serveMetrics}, "benchRPS.csv")
+}
 
-		funcs = append(funcs, funcName)
+// Pull a list of images
+func pullImages(t *testing.T, images []string) string {
+	for _, imageName := range images {
+		resp, _, err := funcPool.Serve(context.Background(), "plr_fnc", imageName, "record")
+		require.NoError(t, err, "Function returned error")
+		require.Equal(t, resp.Payload, "Hello, record_response!")
 	}
 
-	// Boot VMs
-	for i := 0; i < *vmNum; i++ {
+	pidBytes, err := getFirecrackerPid()
+	require.NoError(t, err, "Cannot get firecracker pid")
+	pid := strings.TrimSpace(strings.Split(string(pidBytes), "\n")[0])
+	return pid
+}
+
+// Boot a range of VMs with given images
+func bootVMs(t *testing.T, images []string, startVMID, endVMID int, omittedPid string) {
+	for i := startVMID; i < endVMID; i++ {
 		vmIDString := strconv.Itoa(i)
-		_, err := funcPool.AddInstance(vmIDString, images[funcs[i%len(funcs)]])
+		_, err := funcPool.AddInstance(vmIDString, images[i%len(images)])
 		require.NoError(t, err, "Function returned error")
 	}
 
-	if !*isWithCache && *isColdStart {
-		log.Info("Profile cold start")
-		dropPageCache()
+	if *isBindSocket {
+		log.Debugf("Binding socket 1")
+		err := bindSocket(omittedPid)
+		require.NoError(t, err, "Bind Socket returned error")
 	}
+}
 
-	ticker := time.NewTicker(timeInterval)
-	var vmGroup sync.WaitGroup
-	done := make(chan bool, 1)
+// Inject many requests per second to VMs and profile
```

Comment:
not a precise comment. "many" means how many? be more precise

Reply:
Done